src/klein/_isession.py

```diff
@@ -37,14 +32,96 @@ class TransactionEnded(Exception):
     """
 
 
+class SessionMechanism(Names):
+    """
+    Mechanisms which can be used to identify and authenticate a session.
+
+    @cvar Cookie: The Cookie session mechanism involves looking up the session
+        identifier via an HTTP cookie.  Session objects retrieved via this
+        mechanism may be vulnerable to U{CSRF attacks
+        <https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)>}
+        and therefore must have CSRF protections applied to them.
+
+    @cvar Header: The Header mechanism retrieves the session identifier via a
+        separate header such as C{"X-Auth-Token"}.  Since a different-origin
+        site in a browser can easily send a form submission including cookies,
+        but I{can't} easily put stuff into other arbitrary headers, this does
+        not require additional protections.
+    """
+
+    Cookie = NamedConstant()
+    Header = NamedConstant()
+
+
+class ISession(Interface):
+    """
+    An L{ISession} provider contains an identifier for the session, information
+    about how the session was negotiated with the client software, and
+    """
+
+    identifier = Attribute(
+        """
+        L{unicode} identifying a session.
+
+        This value should be:
+
+            1. I{unique} - no two sessions have the same identifier
+
+            2. I{unpredictable} - no one but the receipient of the session
+               should be able to guess what it is
+
+            3. I{opaque} - it should contain no interesting information
+        """
+    )
+
+    isConfidential = Attribute(
+        """
+        A L{bool} indicating whether this session mechanism transmitted over an
+        encrypted transport, i.e., HTTPS.  If C{True}, this means that this
```

Comment:
Should this be L{True}

Reply:
yup

Reply:
There's a lot of this.  Will save for another PR.

Reply:
#390